### DISCUSSION TITLE
custome url prefix for the API instead of "/"

### DISCUSSION THREAD

**Discussion by Bhanu88** [fastapi/fastapi#7471] [Questions] [answered]
HI, I want to change the base context of the API from "/" to "/myownapi" please let me know how can I do it.

in the flask, I do something like this  

```
from flask import Blueprint, url_for
from flask_restplus import Api
from configuration import config
from configuration.logger import debug, error, warning

blueprint = Blueprint('sentiment', __name__, url_prefix='/myownapi', template_folder='templates')

```

**Comment by dbanty** [comment]
The equivalent of Flask Blueprints are Routers, check out https://fastapi.tiangolo.com/tutorial/bigger-applications/ for more info.

If you want to mount an entire application under a path I believe you need to make it a sub application like in https://fastapi.tiangolo.com/advanced/sub-applications-proxy/ .

Hopefully one of those gets you what you need!

**Comment by Bhanu88** [comment]
Thanks, I am looking for the sub-application proxy option

**Comment by tiangolo** [comment]
Thanks for the help here @dbanty ! :clap: :bow: 

Thanks for reporting back and closing the issue @Bhanu88  :+1:

**Comment by gamesbook** [comment]
> If you want to mount an entire application under a path I believe you need to make it a sub application like in https://fastapi.tiangolo.com/advanced/sub-applications-proxy/ .

This link is broken (as at Dec 2020) - try https://fastapi.tiangolo.com/advanced/behind-a-proxy/

**Comment by m0rtalis** [comment]
> > If you want to mount an entire application under a path I believe you need to make it a sub application like in https://fastapi.tiangolo.com/advanced/sub-applications-proxy/ .
> 
> This link is broken (as at Dec 2020) - try https://fastapi.tiangolo.com/advanced/behind-a-proxy/

Looks like what you need is https://fastapi.tiangolo.com/advanced/sub-applications/

**Comment by rbhanot4739** [comment]
Is there a way to change/add prefix to base app ? For example the base app prefix is "/" which means all of your endpoints will be served from "/". 

```
app = FastAPI()
@app.get("/home")
def home():
    return "Hello World"
```

This makes the endpoint URLs as `/home`, is it possible to add a prefix like `/api` to base app as well just like we do it for router objects so that endpoint URL is `/api/home`. I have looked at `root_path` option but that seems to have a different effect where in your proxy will strip off the `root_path` before sending the request to fastapi but the prefix is never added to api.

**Comment by JLHasson** [comment]
If you take a look at the [bigger applications documentation](https://fastapi.tiangolo.com/tutorial/bigger-applications/#another-module-with-apirouter) linked above you'll see how to add a base prefix.

Here's a minimal example:
```python
from fastapi import APIRouter, FastAPI

router = APIRouter(prefix='/api')

@router.get('/home')
async def home():
    return {'msg': 'hello'}

app = FastAPI()
app.include_router(router)
```

**Reply by elonzh** [reply]
This is not working for `docs_url` or `redoc_url` because they are added to `app.router` when you instantiate FastAPI.

**Comment by rbhanot4739** [comment]
@JLHasson I think the discussion is about adding the prefix to the `app`, not the APIRouter. There can be certain endpoints which are linked directly to app object and you might not want to have a separate router object but still want to serve those endpoints from the same path like `/api`

**Comment by femiayotubosun** [comment]
> @JLHasson I think the discussion is about adding the prefix to the `app`, not the APIRouter. There can be certain endpoints which are linked directly to app object and you might not want to have a separate router object but still want to serve those endpoints from the same path like `/api`

Have you found a solution to this?

**Comment by gamesbook** [comment, marked as answer]
Have you tried with the suggested links - https://fastapi.tiangolo.com/advanced/sub-applications/ and https://fastapi.tiangolo.com/advanced/behind-a-proxy/ including https://fastapi.tiangolo.com/advanced/behind-a-proxy/#mounting-a-sub-application ?

**Comment by styoe** [comment]
You can make a middleware that will rewrite the path just after you call the FastApi constructor like this:
```
app = FastAPI(
    title=config.title, description=config.description, version=config.version,
)

@app.middleware("http")
async def add_root_path(request, call_next):
    request.scope['path'] = "/my-path-prefix" +  request.scope['path']
    return await call_next(request)
```

**Comment by elijahsgh** [comment]
Hi. I'm just chiming in.

I have some load balancers that do not fix up paths (like ingress and loadbalancer pairs on cloud providers). I wasn't able to sort out the APIRouter to enable me to strip part of the path. I have a hand full of applications I want to add API features to. Some are /blog, /blogs, etc and would like to be able to have FastAPI, swagger, etc to reflect paths like <site>/blog/api and <site>/blogs/api

The answer from @styoe does work but it seems now my app is answering both /status and /myprefix/status which is slightly confusing.

Being able to define a path at runtime would be wonderful.

Here is some extra code based on styoe's example (on curl /blogs/api/v1/status I would like /status to answer). `settings.request_path` here is /blogs/api/v1:
```
@app.middleware("http")
async def remove_prefix(request, next):
    rewrite = request.scope['path'].replace(settings.request_prefix, '')

    logger = logging.getLogger('uvicorn.info')
    logger.info(f'Received request, rewriting {rewrite}')

    request.scope['path'] = rewrite
    return await next(request)
```
